core/src/main/java/org/apache/calcite/plan/SubstitutionVisitor.java

```diff
@@ -127,6 +128,11 @@
  */
 public class SubstitutionVisitor {
   private static final boolean DEBUG = CalciteSystemProperty.DEBUG.value();
+  private static final Strong STRONG = new Strong() {
```

Comment:
This deserves a small comment.  why are input refs assumed to be always null?
I guess it's because you will apply it to some specific types of expressions.